erizo/src/erizo/media/codecs/Coder.h

```diff
@@ -0,0 +1,96 @@
+#ifndef ERIZO_SRC_ERIZO_MEDIA_CODECS_CODER_H_
+#define ERIZO_SRC_ERIZO_MEDIA_CODECS_CODER_H_
+
+#include <functional>
+#include <boost/thread.hpp>
+#include <chrono>  // NOLINT
+
+extern "C" {
+#include <libavcodec/avcodec.h>
+#include <libavformat/avformat.h>
+#include <libavutil/frame.h>
+#include <libavutil/audio_fifo.h>
+#include <libavutil/opt.h>
+#include <libavutil/imgutils.h>
+#include <libavutil/error.h>
+#include <libavutil/samplefmt.h>
+#include <libavutil/channel_layout.h>
+#include <libavutil/mathematics.h>
+#include <libavutil/avutil.h>
+#include <libavresample/avresample.h>
+}
+
+#include "./logger.h"
+
+inline static const std::string av_make_error_string_cpp(int errnum) {
+  char errbuf[AV_ERROR_MAX_STRING_SIZE];
+  av_strerror(errnum, errbuf, AV_ERROR_MAX_STRING_SIZE);
+  return (std::string)errbuf;
+}
+
+#define av_err2str_cpp(errnum) av_make_error_string_cpp(errnum).c_str()
+
+namespace erizo {
+
+typedef std::function<void(AVPacket *packet, AVFrame *frame, bool success)> EncodeCB;
+typedef std::function<void(AVCodecContext *context, AVDictionary *dict)> InitContextCB;
+
+enum CoderOperationType {
+  ENCODE_AV,
+  DECODE_AV
+};
+
+class Coder {
+  DECLARE_LOGGER();
+
+ public:
+  bool initContext(AVCodecContext **codec_ctx, AVCodec **av_codec, AVCodecID codec_id,
+      CoderOperationType operation, const InitContextCB callback);
+  bool allocCodecContext(AVCodecContext **ctx, AVCodec **c, AVCodecID codec_id,
+    CoderOperationType operation);
+  bool openCodecContext(AVCodecContext *codec_ctx, AVCodec *codec, AVDictionary *opt);
+  void logCodecContext(AVCodecContext *codec_ctx);
+  bool decode(AVCodecContext *decode_ctx, AVFrame *frame, AVPacket *av_packet);
+  void encode(AVCodecContext *encode_ctx, AVFrame *frame, AVPacket *av_packet,
+      const EncodeCB &done);
+  void saveFrameAsJPEG(AVFrame *frame);
+  static void logAVStream(AVStream *av_stream);
+};
+
+class CoderCodec {
+  DECLARE_LOGGER();
+
+ public:
+  CoderCodec();
+  virtual ~CoderCodec();
+  virtual int closeCodec();
+  void logCodecContext();
+
+ public:
+  bool initialized;
+  AVCodec* av_codec_;
```

Comment:
I think all of these should be protected. In general I wouldn't want to access the pointers directly from outside of the `CoderEncoder` or `CoderDecoder` if possible.

Reply:
Ok, you got it. I'll move `addOutputStream` logic inside CoderCodec and protect all that things.

Reply:
@lodoyun sadly I'll leave accesible `codec_context_` and `frame_` a little bit more. The thing is that hide those almost require a complete layer of abstraction of ffmpeg, which I don't know if it is the intention or not, I have started this abstraction trying to avoid repetition like open/close codecs and context and off-course decoding/encoding logic but still relaying on certain ffmpeg functions. I am not telling which is the right way to go since I didn't clarify on my mind either, but before go through that path I believe it require an specific discussion about it and how to achieve it.

Reply:
Fair enough. We can keep it like this for now.